File: leetcode/139.WordBreak.go
package leetcode

func wordBreak(s string, wordDict []string) bool {
	mappings := make(map[int]bool)
	var traverse func(i int) bool

	traverse = func(i int) bool {
		if b, ok := mappings[i]; ok {
			return b
		}
		if i > len(s)-1 {
			return true
		}

		isMatched := false
		for _, word := range wordDict {
			length := len(word)
			if (i + length) > len(s) {
				continue
			}

			if s[i:i+length] == word {
				nextMatched := traverse(i + length)
				if nextMatched {
					isMatched = true
					break
				}
			}
		}
		mappings[i] = isMatched
		return isMatched
	}

	return traverse(0)
}

File: leetcode/139.WordBreak.ts
function wordBreak(s: string, wordDict: string[]): boolean {
  const map = new Map<number, boolean>()

  const dfs = (i: number): boolean => {
    if (map.has(i)) return map.get(i)!
    if (i > s.length - 1) return true

    let isMatched = false
    for (const word of wordDict) {
      const length = word.length
      if (s.slice(i, i + length) === word) {
        const nextMatched = dfs(i + length)
        if (nextMatched) {
          isMatched = true
          break
        }
      }
    }
    map.set(i, isMatched)
    return isMatched
  }

  return dfs(0)
}
console.log(wordBreak('leetcode', ['leet', 'code']))
